pkg/ingester/downscale.go

```diff
@@ -52,6 +52,14 @@ func (i *Ingester) PrepareInstanceRingDownscaleHandler(w http.ResponseWriter, r
 		i.circuitBreaker.push.deactivate()
 
 	case http.MethodDelete:
+		// Don't leave read-only mode if there's a forced compaction pending or in progress.
+		if len(i.forceCompactTrigger) > 0 || i.forcedCompactionInProgress.Load() {
```

Comment:
I think there are a couple of race conditions here:
- One trivial one is that maybe we've consumed the `forceCompactTrigger` channel but didn't set the `forceCompactionInProgress` mark yet.
- Another tricky one, is the one that I learned the other day that `len(channel)` [is not _really_ atomic](https://stackoverflow.com/a/79021746/3990767).

I think that in order to handle this properly, we need to have another atomic value indicating that `forceCompactTriggered`, increment that before even trying to send to the channel, and decrement that after the force-compaction callback has been returned (or ingester shut down, whatever comes first, i.e., in the defer of that `run()` function).